projects/clr/rocclr/device/rocm/rocvirtual.cpp

```diff
@@ -3736,18 +3737,76 @@ void VirtualGPU::submitMarker(amd::Marker& vcmd) {
 
 // ================================================================================================
 void VirtualGPU::submitAccumulate(amd::AccumulateCommand& vcmd) {
-  // Make sure VirtualGPU has an exclusive access to the resources
   amd::ScopedLock lock(execution());
   profilingBegin(vcmd);
 
+  auto& packets = vcmd.getAccumulatedPackets();
+  const auto& kernelNames = vcmd.getKernelNames();
+  const auto total_packets = vcmd.getPacketCount();
+  const uint32_t queueSize = gpu_queue_->size;
+  const uint32_t queueMask = queueSize - 1;
+  const uint32_t sw_queue_size = queueMask;
+#if 0
+  if (total_packets == 0) {
+    const Settings& settings = dev().settings();
+    if (settings.barrier_value_packet_) {
+      dispatchBarrierValuePacket(kBarrierVendorPacketNopScopeHeader, true);
+    } else {
+      dispatchBarrierPacket(kNopPacketHeader, false);
+    }
+    printf("Empty Accumulate dispatch: SWq=0x%zx, HWq=0x%zx, id=%d, total_packets=%u, "
+           "queue_size=%u, rptr=%u, wptr=%u, doorbell=0x%zx\n",
+           gpu_queue_, gpu_queue_->base_address, gpu_queue_->id, total_packets,
+           gpu_queue_->size,
+           hsa_queue_load_read_index_relaxed(gpu_queue_),
+           hsa_queue_load_write_index_relaxed(gpu_queue_),
+           gpu_queue_->doorbell_signal.handle);
+    profilingEnd();
+    return;
+  }
+#endif
+  uint64_t current_write = hsa_queue_load_write_index_relaxed(gpu_queue_);
+  uint64_t current_read = hsa_queue_load_read_index_relaxed(gpu_queue_);
+  
+  while ((current_write - current_read + total_packets) >= sw_queue_size) {
+    amd::Os::yield();
+    current_write = hsa_queue_load_write_index_relaxed(gpu_queue_);
+    current_read = hsa_queue_load_read_index_scacquire(gpu_queue_);
+  }
+
+  uint64_t start_index = hsa_queue_add_write_index_screlease(gpu_queue_, total_packets);
+
+  for (size_t i = 0; i < total_packets; ++i) {
+    uint64_t packet_index = start_index + i;
+    uint32_t queue_mask = gpu_queue_->size - 1;
+
+    void* queue_slot = static_cast<char*>(gpu_queue_->base_address) +
```

Comment:
the first packet written has to be with invalid packet header else it breaks the CP, it can read invalid data if one hasnt completed writing the body., Also why are you doing a memcpy for a single write